af_ktls.c

```diff
@@ -358,30 +379,569 @@ static void tls_update_senpage_ctx(struct tls_sock *tsk, size_t size)
 	 * can use whole scatterlist next time
 	 */
 	memmove(sg, sg_start,
-		(KTLS_SG_DATA_SIZE - 1 - put_count)*sizeof(tsk->sendpage_ctx.sg[0]));
+		(KTLS_SG_DATA_SIZE - 1 - put_count)*sizeof(
+				tsk->sendpage_ctx.sg[0]));
 	sg_mark_end(&sg[tsk->sendpage_ctx.used]);
 }
 
-#include "dtls-window.c"
+/*Must be called with socket callback locked */
+static void tls_unattach(struct tls_sock *tsk)
+{
+	tsk->socket->sk->sk_data_ready = tsk->saved_sk_data_ready;
+	tsk->socket->sk->sk_user_data = NULL;
+}
+
+static void tls_err_abort(struct tls_sock *tsk)
+{
+	struct sock *sk;
+
+	sk = (struct sock *)tsk;
+	tsk->rx_stopped = 1;
+	sk->sk_err = -EBADMSG;
+	sk->sk_error_report(sk);
+	tsk->saved_sk_data_ready(tsk->socket->sk);
+	tls_unattach(tsk);
+}
+
+static int decrypt_skb(struct tls_sock *tsk, struct sk_buff *skb)
+{
+	int ret, nsg;
+	size_t prepend, overhead;
+	struct tls_rx_msg *rxm;
+
+	prepend = IS_TLS(tsk) ? KTLS_TLS_PREPEND_SIZE : KTLS_DTLS_PREPEND_SIZE;
+	overhead = IS_TLS(tsk) ? KTLS_TLS_OVERHEAD : KTLS_DTLS_OVERHEAD;
+	rxm = tls_rx_msg(skb);
+	/*
+	 * Possible security vulnerability since skb can require
+	 * more than ALG_MAX_PAGES sg's.
+	 * TODO: So what exactly happens if skb_to_sgvec causes more
+	 * than ALG_MAX_PAGES fragments? Consider allocating kernel
+	 * pages
+	 * tls_read_size already copied headers and aad. Therefore
+	 * this simply needs to pass the encrypted data + message
+	 */
+	nsg = skb_to_sgvec(skb, &tsk->sgin[1], rxm->offset +
+			prepend,
+			rxm->full_len - prepend);
+	tls_make_aad(tsk, 1, tsk->aad_recv,
+			rxm->full_len - overhead,
+			tsk->iv_recv);
+
+	/*
+	 * Decrypt in place.
```

Comment:
The alternative would be to allocate another skb for decrypted data. According to Dave, decrypting in place does not incur any extra overhead. The hope is that once we fix high memory issues, we can use skb_clone() instead of skb_copy(), therefore requiring less space for decryption.